File: 2020/02/part1.clj
(require '[clojure.string])

(defn get-input []
  (slurp "/Users/myrsmeden/Development/adventofcode/2020/02/input.txt"))

(defn isPasswordValid [line]
  (let [[policy password] (clojure.string/split line #": ")]
    (let [[numberOfTimes letter] (clojure.string/split policy #" ")]
      (let [[minNumberOfTimes maxNumberOfTimes] (map #(Integer/parseInt %) (clojure.string/split numberOfTimes #"-"))]
        (as-> (reduce (fn [count testLetter]
                        (+ count (if (= testLetter letter) 1 0))) 0 (clojure.string/split password #"")) occurrences
              (<= minNumberOfTimes occurrences maxNumberOfTimes))))))

(count (filter isPasswordValid (clojure.string/split (get-input) #"\n")))

File: 2020/02/part1.js
const fs = require("fs");
const path = require("path");
const input = fs.readFileSync(path.join(__dirname, "input.txt"), {
  encoding: "utf8",
  flag: "r",
});

const passwordPolicies = input.split("\n");

const isPasswordValid = (policy, password) => {
  const [numberOfTimes, letter] = policy.split(" ");
  const [minNumberOfTimes, maxNumberOfTimes] = numberOfTimes
    .split("-")
    .map((num) => parseInt(num, 10));

  const numberOfOccurrences = password
    .split("")
    .reduce((count, pLetter) => (count += pLetter === letter ? 1 : 0), 0);

  return (
    numberOfOccurrences >= minNumberOfTimes &&
    numberOfOccurrences <= maxNumberOfTimes
  );
};
const validPasswords = passwordPolicies.filter((policy) =>
  isPasswordValid(...policy.split(": "))
);

console.log(validPasswords.length);
